
# Figure
```{r, eval=FALSE}
library(ggplot2)

MicroSEC = read.xlsx(paste(wd, "/output/MicroSEC-result_2020-10-31.xlsx", sep=""), sheet = 1)

df = MicroSEC %>% dplyr::filter(Total_read >= 10 & Filter_6_simple_repeat == FALSE & Filter_8_mutation_at_homopolymer == FALSE)
g = ggplot(df, aes(x = Low_quality_base_rate_under_Q18, y = ..density..,))
g = g + geom_histogram(binwidth=0.01, fill = "grey", color = "black")
g = g + theme_set(theme_bw(base_size = 14))
plot(g)
ggsave(paste(wd, "/output/Low_quality_base_rate_under_Q18_", Sys.Date(), ".pdf", sep=""), plot=g, width=5, height=4)

# 5,6,7,8,,,,40
Total_count = rep(0,36)
Passed_count = rep(0,36)
Filtered_rate = rep(0,36)
for(i in 1:36){
  Total_count[i] = dim(MicroSEC %>% dplyr::filter(Total_read == (4 + i) & Filter_6_simple_repeat == FALSE & Filter_8_mutation_at_homopolymer == FALSE))[1]
  Passed_count[i] = dim(MicroSEC %>% dplyr::filter(Total_read == (4 + i) & Filter_6_simple_repeat == FALSE & Filter_8_mutation_at_homopolymer == FALSE & MicroSEC_filter_12345 == FALSE))[1]
  Filtered_rate[i] = 1 - zero(Passed_count[i], Total_count[i])
}

second_rate = max(Total_count) / 100

df = data.frame(
    threshold  = 5:40,
    Total_count = Total_count,
    Passed_count = Passed_count,
    Filtered_rate = Filtered_rate * second_rate * 100
)

g = ggplot(data=df) +
  xlab("Mutation depth") +
  ylab("Mutations") +
  ggtitle("Total mutations, Mutations passing the filter, and filtered rate.") +
  scale_x_continuous(breaks=seq(0,41,10)) +
  scale_y_continuous(breaks=seq(0,4000,500), sec.axis = sec_axis(~ . / second_rate, name = "Filtered rate(%)")) +
  geom_bar(aes(x=threshold, y=Total_count),
           stat="identity", position="identity", 
           alpha=0.2) + 
  geom_bar(aes(x=threshold,y=Passed_count),
            stat="identity", position="identity",
            fill="blue", colour="white", alpha=0.6) +
  geom_line(aes(x=threshold,y=Filtered_rate), 
             stat="identity",
             position="identity",
             colour="black") +
  theme(legend.position="right")
ggsave(paste(wd, "/output/Filtered_rate_", Sys.Date(), ".pdf", sep=""), plot=g, width=10, height=8)

```


# File formatting
```{r, eval=FALSE}
CELL = c("PC9", "RH30", "LS411N", "H3122")
TYPE = c("Cell_line", "FEPE")
for(CELL_ in CELL){
  for(TYPE_ in TYPE){
    files = data.frame(file = list.files(paste(wd, "/source/Cell_line/013_make_vcf_", CELL_, "_", TYPE_, "_Ag_TDv4", sep=""), pattern=".gz", full.names=T))
    files = files %>% dplyr::filter(!str_detect(file, "depth") & !str_detect(file, "tbi") & !str_detect(file, "finish"))
    df_mut_call = NULL
    for(i in files$file){
      df_mut_call  = rbind(df_mut_call, fread(as.character(i), stringsAsFactors=FALSE, header=TRUE, sep="\t") %>% dplyr::filter(Alt != "_"))
    }
    write_tsv(df_mut_call, paste(wd, "/source/", CELL_, "_", TYPE_, "/", CELL_, "_", TYPE_, "_Ag_TDv4.gz", sep=""))
  }
}

CCLE = fread(paste(wd, "/source/CCLE_DepMap_18q3_maf_20180718.txt", sep=""), stringsAsFactors=FALSE, header=TRUE, sep="\t")
CCLE = CCLE %>% dplyr::filter(Tumor_Sample_Barcode %in% c("NCIH3122_LUNG", "RH30_SOFT_TISSUE", "PC9_LUNG", "LS411N_LARGE_INTESTINE"))
CCLE = CCLE %>% dplyr::select(Sample=Tumor_Sample_Barcode, Gene=Hugo_Symbol, HGVS.p=Protein_Change, Chr=Chromosome, Pos=Start_position, Start=Start_position, End=End_position, Strand=Strand, Mut_type=Variant_Type, Ref=Reference_Allele, Alt=Tumor_Seq_Allele1)

CCLE = CCLE %>% dplyr::mutate(Chr=paste("chr", CCLE$Chr, sep=""))
CCLE = CCLE %>% dplyr::mutate(hg19tohg38=paste(CCLE$Chr, ":", CCLE$Start, "-", CCLE$End, sep=""))
write_tsv(data.frame(CCLE$hg19tohg38), file=paste(wd, "/source/CCLE_hg19tohg38.txt", sep=""), col_names=FALSE)

# Lift Genome Annotations
# https://genome.ucsc.edu/cgi-bin/hgLiftOver

hg38_pos = read.csv(paste(wd, "/source/hglft_genome_41a45_6b5600.txt", sep=""), header=FALSE, sep="\t")
CCLE$Chr = str_split(hg38_pos[[1]], ":", simplify = TRUE)[,1]
CCLE$Pos = str_split(str_split(hg38_pos[[1]], ":", simplify = TRUE)[,2], "-", simplify = TRUE)[,1]
CCLE$Start = str_split(str_split(hg38_pos[[1]], ":", simplify = TRUE)[,2], "-", simplify = TRUE)[,1]
CCLE$End = str_split(str_split(hg38_pos[[1]], ":", simplify = TRUE)[,2], "-", simplify = TRUE)[,2]
CCLE = CCLE %>% dplyr::mutate(`Total_QV>=20`=0, `%Alt`=0, SimpleRepeat_TRF="NA", Transition="NA")
trans <- data.frame(
  CELL = c("PC9", "RH30", "LS411N", "H3122"),
  CCLE = c("PC9_LUNG", "RH30_SOFT_TISSUE", "LS411N_LARGE_INTESTINE", "NCIH3122_LUNG")
)
trans = lapply(CCLE$Sample, function(x) {
  as.vector(trans$CELL[match(x, trans$CCLE)])
})
dim(trans) = c(length(trans), 1)
CCLE$Sample = trans
CCLE = CCLE %>% dplyr::mutate(Mut_type = tolower(Mut_type))

fun_genome = function(x, y){
  r = NULL
  for(i in 1:length(x)){
    r = c(r, as.character(genome[[x[i]]][y[i]]))
  }
  return(r)
}

fun_genome_2 = function(x, y, z){
  r = NULL
  for(i in 1:length(x)){
    r = c(r, as.character(genome[[x[i]]][y[i]:z[i]]))
  }
  return(r)
}

CCLE = CCLE %>% dplyr::mutate(PRE_del = fun_genome(Chr, as.integer(Start)-1))
CCLE = CCLE %>% dplyr::mutate(PRE_ins = fun_genome(Chr, as.integer(Start)))
CCLE = CCLE %>% dplyr::mutate(POST_ins = fun_genome(Chr, as.integer(End)))
CCLE = CCLE %>% dplyr::mutate(Mut_type = ifelse(Mut_type == "snp", "snv", Mut_type))
CCLE = CCLE %>% dplyr::mutate(Alt_length_1 = nchar(Ref), Alt_length_2 = nchar(Alt))
CCLE = CCLE %>% dplyr::mutate(Alt_length = (((Alt_length_1 - Alt_length_2) + abs(Alt_length_1 - Alt_length_2)) / 2) + Alt_length_2)

CCLE = CCLE %>% dplyr::mutate(Ref_ins = ifelse(Mut_type == "ins", PRE_ins, ""))
CCLE = CCLE %>% dplyr::mutate(Ref_del = ifelse(Mut_type == "del", paste(PRE_del, Ref, sep=""), ""))
CCLE = CCLE %>% dplyr::mutate(Ref_snv = ifelse(Mut_type == "snv", Ref, ""))

CCLE = CCLE %>% dplyr::mutate(Alt_ins = ifelse(Mut_type == "ins", paste(PRE_ins, Alt, sep=""), ""))
CCLE = CCLE %>% dplyr::mutate(Alt_del = ifelse(Mut_type == "del", PRE_del, ""))
CCLE = CCLE %>% dplyr::mutate(Alt_snv = ifelse(Mut_type == "snv", Alt, ""))

CCLE = CCLE %>% dplyr::mutate(Alt_indel = paste(Alt_ins, Alt_del, Alt_snv, sep=""))
CCLE = CCLE %>% dplyr::mutate(Ref_indel = paste(Ref_ins, Ref_del, Ref_snv, sep=""))

CCLE = CCLE %>% dplyr::mutate(Neighbor_start_1 = as.integer(Start) - 20, Neighbor_end_1 = as.integer(Start) - 1)
CCLE = CCLE %>% dplyr::mutate(Neighbor_start_2 = as.integer(End) + 1, Neighbor_end_2 = as.integer(End) + 20)
CCLE = CCLE %>% dplyr::mutate(Pre_Neighbor = fun_genome_2(Chr, Neighbor_start_1, Neighbor_end_1))
CCLE = CCLE %>% dplyr::mutate(Post_Neighbor = fun_genome_2(Chr, Neighbor_start_2, Neighbor_end_2))
CCLE = CCLE %>% dplyr::mutate(Neighborhood_sequence = ifelse(Mut_type == "ins", paste(Pre_Neighbor, Alt_indel, POST_ins, str_sub(Post_Neighbor, 1, 19), sep=""), ifelse(Mut_type == "del", paste(Pre_Neighbor, Post_Neighbor, sep=""), paste(Pre_Neighbor, Alt, Post_Neighbor, sep=""))))

CCLE = CCLE %>% dplyr::mutate(Mut_type = paste(Alt_length, "-", Mut_type, sep=""))
CCLE = CCLE %>% dplyr::mutate(Ref = Ref_indel, Alt = Alt_indel)
CCLE = CCLE %>% dplyr::select(Sample, Gene, HGVS.p, Mut_type, `Total_QV>=20`, `%Alt`, Chr, Pos, Ref, Alt, SimpleRepeat_TRF, Neighborhood_sequence, Transition)

write.xlsx(as.data.frame(as.matrix(CCLE)), paste(wd, "/source/CCLE.xlsx", sep=""))
```

```{r}
library(devtools)
library(roxygen2)
library(usethis)


```


## Information about the current R session
```{r, eval=FALSE}
sessionInfo()
```
