private source codes

Potential Sequence Error with Uneven Distribution-Omitter
PSEUDO

## Setting
```{r, eval=FALSE}
wd = "/mnt/HDD8TB/MicroSEC"
knitr::opts_chunk$set(collapse = TRUE, fig.width=12, fig.height=8, echo=TRUE, warning=FALSE, message=TRUE, comment = "#>")
knitr::opts_knit$set(root.dir = wd)
options(rmarkdown.html_vignette.check_title = FALSE)
```

## Necessary packages
```{r packages, eval=FALSE}
library(tidyr)
library(openxlsx)
library(data.table)
library(R.utils)
library(stringr)
library(magrittr)
library(dplyr)
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Mmusculus.UCSC.mm10)
library(GenomicAlignments)
library(Rsamtools)
library(tcltk)
library(gtools)
library(MicroSEC)
library(tidyverse)
```

## Sample information file making step
```{r}
# TOP somatic mutations reported to patients
df_mutation_raw = read.xlsx("/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/TOP315_summary.xlsx", sheet = 3)
df_mutation_raw = df_mutation_raw %>% dplyr::select(ID, 遺伝子, DNA変異, アミノ酸変異, `アリル頻度・リード数・コピー数・LRRスウスウ`, マップ率, 平均深度, 総リード数スウ, ユニークリード率)
colnames(df_mutation_raw) = c("Sample_ID", "Gene", "HGVS.c", "HGVS.p", "Freq_info", "Mapped_rate", "Mean_depth", "Total_reads", "Unique_read_rate")
df_mutation_raw = subset(df_mutation_raw, !(is.na(df_mutation_raw$HGVS.c)))
df_mutation_raw$HGVS.p = replace(df_mutation_raw$HGVS.p, which(is.na(df_mutation_raw$HGVS.p)), "_")

# Sample data formatting
df_sample_info = read_tsv("/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/Sample_initial_list.txt")
df_sample_list = NULL
Sample_ID_list = NULL
df_list = NULL
for(SAMPLE in df_sample_info$Sample){
  directory = paste("/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/050_make_mutation_list_", SAMPLE, sep="") 
  file = list.files(directory, pattern=".gz", full.names=T)
  df_sample_list_tmp = read_tsv(as.character(file))
  Genes = (df_mutation_raw %>% dplyr::filter(Sample_ID == df_sample_info[df_sample_info$Sample == SAMPLE,"Sample_ID"][[1]]))$Gene
  df_sample_list_tmp = df_sample_list_tmp %>% dplyr::filter(Gene %in% Genes)
  Sample_ID_list_tmp = unique((df_sample_list_tmp %>% dplyr::mutate(Sample_ID = df_sample_info[df_sample_info$Sample == SAMPLE,"Sample_ID"][[1]]))$Sample_ID)
  Sample_ID_list = c(Sample_ID_list, Sample_ID_list_tmp)
  df_sample_list = rbind(df_sample_list, df_sample_list_tmp)
  
  df_read_summary =  read.csv(paste("/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/051_read_summary_", SAMPLE, "/", SAMPLE, "_RS.txt", sep=""), header=TRUE, stringsAsFactors=FALSE, sep="\t")
  
  mut_file_path = "/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/mutation_info.xlsx"
  bam_file_path = paste("/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/090_merge_bam_", SAMPLE, "/", SAMPLE, ".realigned.bam", sep="")
  ID_directory_path = paste("/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/020_mut_call_", SAMPLE, sep="")
  Read_length_tmp = df_read_summary$Read_length
  Adapter = "AGATCGGAAGAGC"
  Organism = "Human"
  df_list = rbind(df_list, c(SAMPLE, mut_file_path, bam_file_path, ID_directory_path, Read_length_tmp, Adapter, Organism))
}
df_mutation_selected_samples = df_mutation_raw %>% dplyr::filter(Sample_ID  %in% Sample_ID_list)
write.xlsx(df_mutation_selected_samples, "/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/mutation_selected.xlsx")
write.xlsx(df_sample_list, "/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/mutation_info.xlsx")
write.table(df_list, "/mnt/HDD8TB/MicroSEC/source/TOP_MicroSEC_1_20/sample_info.tsv", append = FALSE, quote = FALSE, sep = "\t", row.names = FALSE,col.names = FALSE)


```

# Figure
```{r, eval=FALSE}
library(ggplot2)

MSEC_result = read.xlsx("/mnt/HDD8TB/MicroSEC/MicroSEC-result_CAUP-H17-08384-1-T_NS_TDv3_2020-11-05.xlsx", sheet = 1)

df = MSEC_result %>% dplyr::filter(Total_read >= 10 & Filter_6_simple_repeat == FALSE & Filter_8_mutation_at_homopolymer == FALSE)
g = ggplot(df, aes(x = Low_quality_base_rate_under_Q18, y = ..density..,))
g = g + geom_histogram(binwidth=0.01, fill = "grey", color = "black")
g = g + theme_set(theme_bw(base_size = 14))
plot(g)
ggsave(paste(wd, "/output/Low_quality_base_rate_under_Q18_", Sys.Date(), ".pdf", sep=""), plot=g, width=5, height=4)

# 5,6,7,8,,,,40
Total_count = rep(0,36)
Passed_count = rep(0,36)
Filtered_rate = rep(0,36)
for(i in 1:36){
  Total_count[i] = dim(MSEC_result %>% dplyr::filter(Total_read == (4 + i) & Filter_6_simple_repeat == FALSE & Filter_8_mutation_at_homopolymer == FALSE))[1]
  Passed_count[i] = dim(MSEC_result %>% dplyr::filter(Total_read == (4 + i) & Filter_6_simple_repeat == FALSE & Filter_8_mutation_at_homopolymer == FALSE & MSEC_result_filter_12345 == FALSE))[1]
  Filtered_rate[i] = 1 - zero(Passed_count[i], Total_count[i])
}

second_rate = max(Total_count) / 100

df = data.frame(
    threshold  = 5:40,
    Total_count = Total_count,
    Passed_count = Passed_count,
    Filtered_rate = Filtered_rate * second_rate * 100
)

g = ggplot(data=df) +
  xlab("Mutation depth") +
  ylab("Mutations") +
  ggtitle("Total mutations, Mutations passing the filter, and filtered rate.") +
  scale_x_continuous(breaks=seq(0,41,10)) +
  scale_y_continuous(breaks=seq(0,4000,500), sec.axis = sec_axis(~ . / second_rate, name = "Filtered rate(%)")) +
  geom_bar(aes(x=threshold, y=Total_count),
           stat="identity", position="identity", 
           alpha=0.2) + 
  geom_bar(aes(x=threshold,y=Passed_count),
            stat="identity", position="identity",
            fill="blue", colour="white", alpha=0.6) +
  geom_line(aes(x=threshold,y=Filtered_rate), 
             stat="identity",
             position="identity",
             colour="black") +
  theme(legend.position="right")
ggsave(paste(wd, "/output/Filtered_rate_", Sys.Date(), ".pdf", sep=""), plot=g, width=10, height=8)

```


# File formatting
```{r, eval=FALSE}
CELL = c("PC9", "RH30", "LS411N", "H3122")
TYPE = c("Cell_line", "FEPE")
for(CELL_ in CELL){
  for(TYPE_ in TYPE){
    files = data.frame(file = list.files(paste(wd, "/source/Cell_line/013_make_vcf_", CELL_, "_", TYPE_, "_Ag_TDv4", sep=""), pattern=".gz", full.names=T))
    files = files %>% dplyr::filter(!str_detect(file, "depth") & !str_detect(file, "tbi") & !str_detect(file, "finish"))
    df_mut_call = NULL
    for(i in files$file){
      df_mut_call  = rbind(df_mut_call, fread(as.character(i), stringsAsFactors=FALSE, header=TRUE, sep="\t") %>% dplyr::filter(Alt != "_"))
    }
    write_tsv(df_mut_call, paste(wd, "/source/", CELL_, "_", TYPE_, "/", CELL_, "_", TYPE_, "_Ag_TDv4.gz", sep=""))
  }
}

CCLE = fread(paste(wd, "/source/CCLE_DepMap_18q3_maf_20180718.txt", sep=""), stringsAsFactors=FALSE, header=TRUE, sep="\t")
CCLE = CCLE %>% dplyr::filter(Tumor_Sample_Barcode %in% c("NCIH3122_LUNG", "RH30_SOFT_TISSUE", "PC9_LUNG", "LS411N_LARGE_INTESTINE"))
CCLE = CCLE %>% dplyr::select(Sample=Tumor_Sample_Barcode, Gene=Hugo_Symbol, HGVS.c=cDNA_Change, HGVS.p=Protein_Change, Chr=Chromosome, Pos=Start_position, Start=Start_position, End=End_position, Strand=Strand, Mut_type=Variant_Type, Ref=Reference_Allele, Alt=Tumor_Seq_Allele1)
CCLE = CCLE %>% dplyr::mutate(`Total_QV>=20`=0, `%Alt`=0, SimpleRepeat_TRF=NA, Neighborhood_sequence="", Transition=NA)

CCLE = CCLE %>% dplyr::mutate(Chr=paste("chr", CCLE$Chr, sep=""))
CCLE = CCLE %>% dplyr::mutate(hg19tohg38=paste(CCLE$Chr, ":", CCLE$Start, "-", CCLE$End, sep=""))
write_tsv(data.frame(CCLE$hg19tohg38), file=paste(wd, "/source/CCLE_hg19tohg38.txt", sep=""), col_names=FALSE)

# Lift Genome Annotations
# https://genome.ucsc.edu/cgi-bin/hgLiftOver

hg38_pos = read.csv(paste(wd, "/source/hglft_genome_41a45_6b5600.txt", sep=""), header=FALSE, sep="\t")
CCLE$Chr = str_split(hg38_pos[[1]], ":", simplify = TRUE)[,1]
CCLE$Pos = str_split(str_split(hg38_pos[[1]], ":", simplify = TRUE)[,2], "-", simplify = TRUE)[,1]
CCLE$Start = str_split(str_split(hg38_pos[[1]], ":", simplify = TRUE)[,2], "-", simplify = TRUE)[,1]
CCLE$End = str_split(str_split(hg38_pos[[1]], ":", simplify = TRUE)[,2], "-", simplify = TRUE)[,2]
CCLE = CCLE %>% dplyr::mutate(`Total_QV>=20`=0, `%Alt`=0, SimpleRepeat_TRF="NA", Transition="NA")
trans <- data.frame(
  CELL = c("PC9", "RH30", "LS411N", "H3122"),
  CCLE = c("PC9_LUNG", "RH30_SOFT_TISSUE", "LS411N_LARGE_INTESTINE", "NCIH3122_LUNG")
)
trans = lapply(CCLE$Sample, function(x) {
  as.vector(trans$CELL[match(x, trans$CCLE)])
})
dim(trans) = c(length(trans), 1)
CCLE$Sample = trans
CCLE = CCLE %>% dplyr::mutate(Mut_type = tolower(Mut_type))

fun_genome = function(x, y){
  r = NULL
  for(i in 1:length(x)){
    r = c(r, as.character(genome[[x[i]]][y[i]]))
  }
  return(r)
}

fun_genome_2 = function(x, y, z){
  r = NULL
  for(i in 1:length(x)){
    r = c(r, as.character(genome[[x[i]]][y[i]:z[i]]))
  }
  return(r)
}

CCLE = CCLE %>% dplyr::mutate(PRE_del = fun_genome(Chr, as.integer(Start)-1))
CCLE = CCLE %>% dplyr::mutate(PRE_ins = fun_genome(Chr, as.integer(Start)))
CCLE = CCLE %>% dplyr::mutate(POST_ins = fun_genome(Chr, as.integer(End)))
CCLE = CCLE %>% dplyr::mutate(Mut_type = ifelse(Mut_type == "snp", "snv", Mut_type))
CCLE = CCLE %>% dplyr::mutate(Alt_length_1 = nchar(Ref), Alt_length_2 = nchar(Alt))
CCLE = CCLE %>% dplyr::mutate(Alt_length = (((Alt_length_1 - Alt_length_2) + abs(Alt_length_1 - Alt_length_2)) / 2) + Alt_length_2)

CCLE = CCLE %>% dplyr::mutate(Ref_ins = ifelse(Mut_type == "ins", PRE_ins, ""))
CCLE = CCLE %>% dplyr::mutate(Ref_del = ifelse(Mut_type == "del", paste(PRE_del, Ref, sep=""), ""))
CCLE = CCLE %>% dplyr::mutate(Ref_snv = ifelse(Mut_type == "snv", Ref, ""))

CCLE = CCLE %>% dplyr::mutate(Alt_ins = ifelse(Mut_type == "ins", paste(PRE_ins, Alt, sep=""), ""))
CCLE = CCLE %>% dplyr::mutate(Alt_del = ifelse(Mut_type == "del", PRE_del, ""))
CCLE = CCLE %>% dplyr::mutate(Alt_snv = ifelse(Mut_type == "snv", Alt, ""))

CCLE = CCLE %>% dplyr::mutate(Alt_indel = paste(Alt_ins, Alt_del, Alt_snv, sep=""))
CCLE = CCLE %>% dplyr::mutate(Ref_indel = paste(Ref_ins, Ref_del, Ref_snv, sep=""))

CCLE = CCLE %>% dplyr::mutate(Neighbor_start_1 = as.integer(Start) - 20, Neighbor_end_1 = as.integer(Start) - 1)
CCLE = CCLE %>% dplyr::mutate(Neighbor_start_2 = as.integer(End) + 1, Neighbor_end_2 = as.integer(End) + 20)
CCLE = CCLE %>% dplyr::mutate(Pre_Neighbor = fun_genome_2(Chr, Neighbor_start_1, Neighbor_end_1))
CCLE = CCLE %>% dplyr::mutate(Post_Neighbor = fun_genome_2(Chr, Neighbor_start_2, Neighbor_end_2))
CCLE = CCLE %>% dplyr::mutate(Neighborhood_sequence = ifelse(Mut_type == "ins", paste(Pre_Neighbor, Alt_indel, POST_ins, str_sub(Post_Neighbor, 1, 19), sep=""), ifelse(Mut_type == "del", paste(Pre_Neighbor, Post_Neighbor, sep=""), paste(Pre_Neighbor, Alt, Post_Neighbor, sep=""))))

CCLE = CCLE %>% dplyr::mutate(Mut_type = paste(Alt_length, "-", Mut_type, sep=""))
CCLE = CCLE %>% dplyr::mutate(Ref = Ref_indel, Alt = Alt_indel)
CCLE = CCLE %>% dplyr::select(Sample, Gene, HGVS.c, HGVS.p, Mut_type, `Total_QV>=20`, `%Alt`, Chr, Pos, Ref, Alt, SimpleRepeat_TRF, Neighborhood_sequence, Transition)

write.xlsx(as.data.frame(as.matrix(CCLE)), paste(wd, "/source/CCLE.xlsx", sep=""))
```

## Sample data generation
```{r}
ID_data = read.csv("/mnt/HDD8TB/MicroSEC/source/extract.txt", header=FALSE, stringsAsFactors=FALSE, sep="\t")
colnames(ID_data) = c("chr", "pos")
ID = rep(FALSE, length(df_BAM$rname))
for(i in 1:dim(ID_data)[1]){
  ID = ID | (df_BAM$rname == ID_data$chr[[i]] &  df_BAM$pos > ID_data$pos[[i]] - 160 & df_BAM$pos < ID_data$pos[[i]] + 10)
}
df_BAM$rname = df_BAM$rname[ID]
df_BAM$qname = df_BAM$qname[ID]
df_BAM$seq = df_BAM$seq[ID]
df_BAM$strand = df_BAM$strand[ID]
df_BAM$cigar = df_BAM$cigar[ID]
df_BAM$qual = df_BAM$qual[ID]
df_BAM$pos = df_BAM$pos[ID]

df_mut_call = df_mut_call %>% 
        filter((Chr == ID_data$chr[[1]] & Pos == ID_data$pos[[1]]) | 
                 (Chr == ID_data$chr[[2]] & Pos == ID_data$pos[[2]]) | 
                 (Chr == ID_data$chr[[3]] & Pos == ID_data$pos[[3]]) |
                 (Chr == ID_data$chr[[4]] & Pos == ID_data$pos[[4]]))
mut_read = df_mut_call[1,]
mut_read_ID_list = str_split(mut_read$Mut_ID, pattern = ";")[[1]]
mut_read_ID_list = mut_read_ID_list[[2]]
mut_read_ID = str_sub(str_split(mut_read_ID_list,",")[[1]], start=1, end=-2)

mut_read = df_mut_call[2,]
mut_read_ID_list = str_split(mut_read$Mut_ID, pattern = ";")[[1]]
mut_read_ID_list = mut_read_ID_list[[2]]
mut_read_ID = c(mut_read_ID, str_sub(str_split(mut_read_ID_list,",")[[1]], start=1, end=-2))

mut_read = df_mut_call[3,]
mut_read_ID_list = str_split(mut_read$Mut_ID, pattern = ";")[[1]]
mut_read_ID_list = mut_read_ID_list[[2]]
mut_read_ID = c(mut_read_ID, str_sub(str_split(mut_read_ID_list,",")[[1]], start=1, end=-2))

mut_read = df_mut_call[4,]
mut_read_ID_list = str_split(mut_read$Mut_ID, pattern = ";")[[1]]
mut_read_ID_list = mut_read_ID_list[[3]]
mut_read_ID = c(mut_read_ID, str_sub(str_split(mut_read_ID_list,",")[[1]], start=1, end=-2))

ID_No = df_BAM$qname %in% mut_read_ID
df_BAM$rname = df_BAM$rname[ID_No]
df_BAM$qname = df_BAM$qname[ID_No]
df_BAM$seq = df_BAM$seq[ID_No]
df_BAM$strand = df_BAM$strand[ID_No]
df_BAM$cigar = df_BAM$cigar[ID_No]
df_BAM$qual = df_BAM$qual[ID_No]
df_BAM$pos = df_BAM$pos[ID_No]

exampleMutation = df_mutation[c(9,10,34,35),]

exampleBAM=NULL
exampleBAM$rname = as.character(df_BAM$rname)
exampleBAM$qname = as.character(df_BAM$qname)
exampleBAM$seq = DNAStringSet(as.character(df_BAM$seq))
exampleBAM$strand = as.character(df_BAM$strand)
exampleBAM$cigar = as.character(df_BAM$cigar)
exampleBAM$qual = as.character(df_BAM$qual)
exampleBAM$pos = as.integer(df_BAM$pos)
exampleMutCall = df_mut_call

usethis::use_data(exampleBAM, overwrite=TRUE)
usethis::use_data(exampleMutCall, overwrite=TRUE)
usethis::use_data(exampleMutation, overwrite=TRUE)

```

## Information about the current R session
```{r, eval=FALSE}
sessionInfo()
```
