{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c55889c-424b-43d0-9e07-7e7022015301",
   "metadata": {},
   "outputs": [],
   "source": [
    "# private source codes\n",
    "\n",
    "## Setting\n",
    "wd = \"/Users/ikegami/MicroSEC\"\n",
    "knitr::opts_chunk$set(collapse = TRUE, fig.width=12, fig.height=8, echo=TRUE, warning=TRUE, message=TRUE, comment = \"#>\")\n",
    "knitr::opts_knit$set(root.dir = wd)\n",
    "options(rmarkdown.html_vignette.check_title = FALSE)\n",
    "fun_zero <- function(a, b)ifelse(b == 0, 0, a / b)\n",
    "setwd(wd)\n",
    "\n",
    "## Necessary packages\n",
    "library(tidyr)\n",
    "library(openxlsx)\n",
    "library(data.table)\n",
    "library(R.utils)\n",
    "library(stringr)\n",
    "library(magrittr)\n",
    "library(dplyr)\n",
    "library(Biostrings)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(GenomicAlignments)\n",
    "library(Rsamtools)\n",
    "library(tcltk)\n",
    "library(gtools)\n",
    "library(MicroSEC)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ba091e5a-96b2-4c9a-975b-a0f2b8ee6965",
   "metadata": {},
   "outputs": [],
   "source": [
    "# target sequencing\n",
    "target_info = read.csv(\"source/target_info.txt\", header=TRUE, stringsAsFactors=FALSE, sep=\"\\t\")\n",
    "target_info$Sample_gene = str_replace_all(target_info$Sample_gene, pattern=\"-\", replacement=\"_\")\n",
    "\n",
    "files_R1 = data.frame(file = list.files(paste(wd, \"/source/validation\", sep = \"\"), pattern=\"R1_001.fastq\", full.names=T))\n",
    "files_R2 = data.frame(file = list.files(paste(wd, \"/source/validation\", sep = \"\"), pattern=\"R2_001.fastq\", full.names=T))\n",
    "files_R1_2 = data.frame(file = list.files(paste(wd, \"/source/validation_2\", sep = \"\"), pattern=\"R1_001.fastq\", full.names=T))\n",
    "files_R2_2 = data.frame(file = list.files(paste(wd, \"/source/validation_2\", sep = \"\"), pattern=\"R2_001.fastq\", full.names=T))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9841e15-3117-4fe5-b52b-51c79c0abd50",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in 1:length(files_R1$file)){\n",
    "  Sample_name = as.character(files_R1$file[i])\n",
    "  Sample_name = str_split(str_split(Sample_name, pattern = \"/\")[[1]][7], \"_S\")[[1]][1]\n",
    "  Sample_name = str_replace_all(Sample_name, pattern = \"-\", replacement = \"_\")\n",
    "  info_tmp = target_info %>% dplyr::filter(Sample_gene == Sample_name)\n",
    "  forward = nchar(info_tmp$forward)\n",
    "  reverse = nchar(info_tmp$reverse)\n",
    "  Alt = nchar(info_tmp$Alt_all)\n",
    "  Ref = nchar(info_tmp$Ref_all)\n",
    "  seq_alt_f = DNAString(info_tmp$Alt_all)[(forward - 3):(Alt - reverse + 4)]\n",
    "  seq_alt_r = reverseComplement(seq_alt_f)\n",
    "  seq_ref_f = DNAString(info_tmp$Ref_all)[(forward - 3):(Ref - reverse + 4)]\n",
    "  seq_ref_r = reverseComplement(seq_ref_f)\n",
    "  df_fastq = c(readDNAStringSet(as.character(files_R1$file[i]), format=\"fastq\"),  readDNAStringSet(as.character(files_R2$file[i]), format=\"fastq\"))\n",
    "  print(paste(Sample_name, \"ref:\", sum(vcountPattern(pattern = seq_ref_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_ref_r, subject = df_fastq) > 0),\n",
    "                \"alt:\", sum(vcountPattern(pattern = seq_alt_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_alt_r, subject = df_fastq) > 0)))\n",
    "  target_info[target_info$Sample_gene == Sample_name, \"Ref_no\"] = sum(vcountPattern(pattern = seq_ref_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_ref_r, subject = df_fastq) > 0)\n",
    "  target_info[target_info$Sample_gene == Sample_name, \"Alt_no\"] = sum(vcountPattern(pattern = seq_alt_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_alt_r, subject = df_fastq) > 0)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "941cfa75-a614-4de8-8b17-29e68ad7f24f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in 1:length(files_R1_2$file)){\n",
    "  Sample_name = as.character(files_R1_2$file[i])\n",
    "  Sample_name = str_split(str_split(Sample_name, pattern = \"/\")[[1]][7], \"_GAGA\")[[1]][1]\n",
    "  Sample_name = str_split(Sample_name, \"_CTGA\")[[1]][1]\n",
    "  Sample_name = str_split(Sample_name, \"_ATTC\")[[1]][1]\n",
    "  Sample_name = str_split(Sample_name, \"_GAAT\")[[1]][1]\n",
    "  Sample_name = str_split(Sample_name, \"_TAAT\")[[1]][1]\n",
    "  Sample_name = str_replace_all(Sample_name, pattern = \"-\", replacement = \"_\")\n",
    "  info_tmp = target_info %>% dplyr::filter(Sample_gene == Sample_name)\n",
    "  forward = nchar(info_tmp$forward)\n",
    "  reverse = nchar(info_tmp$reverse)\n",
    "  Alt = nchar(info_tmp$Alt_all)\n",
    "  Ref = nchar(info_tmp$Ref_all)\n",
    "  seq_alt_f = DNAString(info_tmp$Alt_all)[(forward - 3):(Alt - reverse + 4)]\n",
    "  seq_alt_r = reverseComplement(seq_alt_f)\n",
    "  seq_ref_f = DNAString(info_tmp$Ref_all)[(forward - 3):(Ref - reverse + 4)]\n",
    "  seq_ref_r = reverseComplement(seq_ref_f)\n",
    "  df_fastq = c(readDNAStringSet(as.character(files_R1_2$file[i]), format=\"fastq\"),  readDNAStringSet(as.character(files_R2_2$file[i]), format=\"fastq\"))\n",
    "  print(paste(Sample_name, \"ref:\", sum(vcountPattern(pattern = seq_ref_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_ref_r, subject = df_fastq) > 0),\n",
    "                \"alt:\", sum(vcountPattern(pattern = seq_alt_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_alt_r, subject = df_fastq) > 0)))\n",
    "  target_info[target_info$Sample_gene == Sample_name, \"Ref_no\"] = sum(vcountPattern(pattern = seq_ref_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_ref_r, subject = df_fastq) > 0)\n",
    "  target_info[target_info$Sample_gene == Sample_name, \"Alt_no\"] = sum(vcountPattern(pattern = seq_alt_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_alt_r, subject = df_fastq) > 0)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a804c8ca-f3cc-4d2e-8437-2733fd484cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "utils::write.table(target_info, \"output/target_seq_result.tsv\", sep = \"\\t\", na = \"\",\n",
    "              row.names = FALSE, col.names = TRUE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0be8e21b-dc18-4f42-865c-944905e95671",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73b9a251-c8b8-4540-aaa1-7eb321ba26f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# target sequencing\n",
    "target_info = read.xlsx(\"source/SL_validation_revise.xlsx\", sheet = 1)\n",
    "target_info$Sample_gene = str_replace_all(target_info$Sample_gene, pattern=\"-\", replacement=\"_\")\n",
    "\n",
    "files_R1 = data.frame(file = list.files(paste(wd, \"/source/validation_revise\", sep = \"\"), pattern=\"R1_001.fastq\", full.names=T))\n",
    "files_R2 = data.frame(file = list.files(paste(wd, \"/source/validation_revise\", sep = \"\"), pattern=\"R2_001.fastq\", full.names=T))\n",
    "\n",
    "for(i in 1:length(files_R1$file)){\n",
    "  Sample_name = as.character(files_R1$file[i])\n",
    "  Sample_name = str_split(str_split(Sample_name, pattern = \"/\")[[1]][7], \"_S\")[[1]][1]\n",
    "  Sample_name = str_replace_all(Sample_name, pattern = \"-\", replacement = \"_\")\n",
    "  info_tmp = target_info %>% dplyr::filter(Sample_gene == Sample_name)\n",
    "  forward = nchar(info_tmp$forward)\n",
    "  reverse = nchar(info_tmp$reverse)\n",
    "  Alt = nchar(info_tmp$Alt_all)\n",
    "  Ref = nchar(info_tmp$Ref_all)\n",
    "  seq_alt_f = DNAString(info_tmp$Alt_all)[(forward - 3):(Alt - reverse + 4)]\n",
    "  seq_alt_r = reverseComplement(seq_alt_f)\n",
    "  seq_ref_f = DNAString(info_tmp$Ref_all)[(forward - 3):(Ref - reverse + 4)]\n",
    "  seq_ref_r = reverseComplement(seq_ref_f)\n",
    "  df_fastq = c(readDNAStringSet(as.character(files_R1$file[i]), format=\"fastq\"),  readDNAStringSet(as.character(files_R2$file[i]), format=\"fastq\"))\n",
    "  print(paste(Sample_name, \"ref:\", sum(vcountPattern(pattern = seq_ref_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_ref_r, subject = df_fastq) > 0),\n",
    "                \"alt:\", sum(vcountPattern(pattern = seq_alt_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_alt_r, subject = df_fastq) > 0)))\n",
    "  target_info[target_info$Sample_gene == Sample_name, \"Ref_no\"] = sum(vcountPattern(pattern = seq_ref_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_ref_r, subject = df_fastq) > 0)\n",
    "  target_info[target_info$Sample_gene == Sample_name, \"Alt_no\"] = sum(vcountPattern(pattern = seq_alt_f, subject = df_fastq) > 0) +\n",
    "                  sum(vcountPattern(pattern = seq_alt_r, subject = df_fastq) > 0)\n",
    "}\n",
    "\n",
    "utils::write.table(target_info, \"output/target_seq_result_revise.tsv\", sep = \"\\t\", na = \"\",\n",
    "              row.names = FALSE, col.names = TRUE, quote = FALSE)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
