---
title: "MicroSEC: Sequence artifact filtering pipeline for FFPE samples"
author: "Masachika Ikegami"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

## Introduction
This pipeline is designed for filtering mutations found in formalin-fixed and 
paraffin-embedded (FFPE) samples. The MicroSEC filter utilizes a statistical 
analysis, and the results for mutations with less than 10 supporting reads are 
not reliable. Three files are nessesary for the analysis: mutation information 
file and BAM file, and two files optional.  
  
File 1: mutation information tsv file (mandatory)  
This tsv file with column names should contain at least these contents:  
       Sample Mut_type  Chr        Pos Ref Alt SimpleRepeat_TRF
       Neighborhood_sequence  
 SL_1010-N6-B    1-snv chr1  108130741   C   T                N
 CTACCTGGAGAATGGGCCCATGTGTCCAGGTAGCAGTAAGC  
  
SimpleRepeat_TRF: Whether the mutation locates at a simple repeat sequence or 
not (please set "Y" or "N").  
Neighborhood_sequence: [5'-20 bases] + [Alt sequence] + [3'-20 bases].  
Sample, Chr, Pos, Ref, and Alt should be set exactly.  
If you do not know the SimpleRepeat_TRF, Mut_type, or Neighborhood_sequence, 
enter "-". Automatically detected.  
  
File 2: BAM file (mandatory)  
File 3: sample information tsv file (mandatory)  
Seven to nine columns are necessary.  
[sample name] [mutation information excel/tsv file] [BAM file] [read length] 
[adapter sequence read 1] [optional: adapter sequence read 2] 
[sample type: Human or Mouse] [panel name] 
[optional: reference genome fastq file] 
[optional: simple repeat region bed file]  
PC9	./source/CCLE.xlsx ./source/Cell_line/PC9_Cell_line_Ag_TDv4.realigned.bam 
./source/PC9_Cell_line 127 
AGATCGGAAGAGCACACGTCTGAACTCCAGTCA AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT hg38 
TOP ./reference/hg38.fa ./reference/simpleRepeat.bed.gz  
  
File 4: Reference genome: Human (hg38 or hg19) or Mouse (mm10) 
(optional, but mandatory with cram files)  

File 5: simple repeat region bed file 
(optional, but mandatory to detect simple repeat derived artifacts)  
  
This pipeline contains 8 filtering processes.  
  
Filter 1  : Shorter-supporting lengths distribute too short to occur 
(1-1 and 1-2).  
Filter 1-1: P-values are less than the threshold_p(default: 10 ^ (-6)).  
Filter 1-2: The longest shorter-supporting lengths is shorter than 40% of the 
read length.  
Filter 2  : Hairpin-structure induced error detection (2-1 or 2-2).  
Filter 2-1: Palindromic sequences exist within 150 bases. 
Filter 2-2: >=50% mutation-supporting reads contains a reverse complementary 
sequence of the opposite strand consisting >= 15 bases.  
Filter 3  : 3’-/5’-supporting lengths are too densely distributed to occur 
(3-1, 3-2, and 3-3).  
Filter 3-1: P-values are less than the threshold_p(default: 10^(-6)).  
Filter 3-2: The distributions of 3’-/5’-supporting lengths are shorter than 80% 
of the read length.  
Filter 3-3: <10% of bases are low quality (Quality score <18).  
Filter 4  : >=15% mutations were called by chimeric reads comprising two 
distant regions.  
Filter 5  : >=50% mutations were called by soft-clipped reads.
Filter 6  : Mutations locating at simple repeat sequences.
Filter 7  : Indel mutations locating at a >=15 homopolymer.  
Filter 8  : >=10% low quality bases in the mutation supporting reads.  

Supporting lengths are adjusted considering small repeat sequences around the 
mutations.  
  
Results are saved in a tsv file.  

github url: https://github.com/MANO-B/MicroSEC  
  
## Setting
```{r}
wd <- "."
knitr::opts_chunk$set(collapse = TRUE,
                      fig.width = 12,
                      fig.height = 8,
                      echo = TRUE,
                      warning = FALSE,
                      message = TRUE,
                      comment = "#>")
options(rmarkdown.html_vignette.check_title = FALSE,
        show.error.messages = FALSE,
        warn = -1)

progress_bar <- "N"
```

## Necessary packages
```{r packages}
library(MicroSEC)
```

## Analysis
```{r}
# please use MicroSEC.R script file which can be downloaded from github website.

# initialize
msec <- NULL
homology_search <- NULL
mut_depth <- NULL

# test data
sample_name <- "H15-11943-1-T_TDv3"
read_length <- 151
adapter_1 <- "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA"
adapter_2 <- "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"
organism <- "hg38"

# load genomic sequence
ref_genome <- fun_load_genome(organism)
chr_no <- fun_load_chr_no(organism)

# load mutation information
df_mutation <- fun_load_mutation(
  system.file("extdata", "test_mutation.tsv", package = "MicroSEC"),
  "H15-11943-1-T_TDv3",
  ref_genome
)

# load bam file (cram file with reference file acceptable)
df_bam <- fun_load_bam(
 system.file("extdata", "test.bam", package = "MicroSEC")
)

# analysis
result = fun_read_check(df_mutation = exampleMutation,
                        df_bam = exampleBAM,
                        df_mut_call = exampleMutCall,
                        ref_genome = ref_genome,
                        sample_name = sample_name,
                        read_length = read_length,
                        adapter_1 = adapter_1,
                        adapter_2 = adapter_2,
                        short_homology_search_length = 4,
                        progress_bar = progress_bar)
msec <- rbind(msec, result[[1]])
homology_search <- rbind(homology_search, result[[2]])
mut_depth <- rbind(mut_depth, result[[3]])

# search homologous sequences
msec <- fun_homology(msec,
                     homology_search,
                     min_homology_search = 40,
                     ref_genome,
                     chr_no,
                     progress_bar = progress_bar)

# statistical analysis
msec <- fun_summary(msec)
msec <- fun_analysis(msec,
                    mut_depth,
                    short_homology_search_length = 4,
                    min_homology_search = 40,
                    threshold_p = 10 ^ (-6),
                    threshold_hairpin_ratio = 0.50,
                    threshold_short_length = 0.75,
                    threshold_distant_homology = 0.15,
                    threshold_soft_clip_ratio = 0.50,
                    threshold_low_quality_rate = 0.1,
                    homopolymer_length = 15)

# save the results in the working/output directory.
# fun_save_gz(msec, "./test_data.tsv")
```

## Results
```{r}
msec
```

## Information about the current R session
```{r}
sessionInfo()
```

