---
title: "MicroSEC: Sequence artifact filtering pipeline for FFPE samples"
author: "Masachika Ikegami"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

## Introduction
This pipeline is designed for filtering mutations found in formalin-fixed and 
paraffin-embedded (FFPE) samples. The MicroSEC filter utilizes a statistical 
analysis, and the results for mutations with less than 10 supporting reads are 
not reliable. Four files are nessesary for the analysis: mutation information 
file, BAM file, and mutation supporting read ID information file.  
  
File 1: mutation information file  
This excel file should contain at least these contents:  
  
       Sample     Gene HGVS.c  HGVS.p Mut_type Total_QV>=20   %Alt  Chr  
 SL_1010-N6-B SLC25A24      _       _    1-snv          366 1.0929 chr1  
  
      Pos Ref Alt SimpleRepeat_TRF                     Neighborhood_sequence  
108130741   C   T                N CTACCTGGAGAATGGGCCCATGTGTCCAGGTAGCAGTAAGC  

Transition  
     C>T_t  
  
Total_QV>=20: The read number with total Q-value >=20.  
SimpleRepeat_TRF: Whether the mutation locates at a simple repeat sequence or 
not.  
Neighborhood_sequence: [5'-20 bases] + [Alt sequence] + [3'-20 bases].  
Transition: 1-snv mutation pattern with a 3'-base. C>T_t represents CT to TT 
mutation. C>T_g_FFPE represents the possible FFPE artifact.  
Sample, Mut_type, Chr, Pos, Ref, and Alt should be set exactly.  
Gene, HGVS.c, HGVS.p, Total_QV>=20, %Alt, SimpleRepeat_TRF, and Transition can 
be set to any values.  
  
File 2: BAM file  
  
File 3: mutation supporting read ID information file  
This file should contain at least these contents:  
  
 Chr     Pos Ref Alt  
chr1 2561609   T   A  
  
                                                             Mut_ID     Mut  
_;ID001-1:579185f,ID004-1:1873933f;ID006-1:1131647f,ID001-1:570086f  .;A;N#  
  
File 4: sample information tsv file  
Seven or eight columns are necessary.  
[sample name] [mutation information excel file] [BAM file] 
[read ID information directory] [read length] [adapter sequence read 1] 
[optional: adapter sequence read 2] [sample type: Human or Mouse]  
PC9	./source/CCLE.xlsx ./source/Cell_line/PC9_Cell_line_Ag_TDv4.realigned.bam 
./source/PC9_Cell_line 127 
AGATCGGAAGAGCACACGTCTGAACTCCAGTCA AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT hg38  
  
This pipeline contains 8 filtering processes.  
  
Filter 1  : Shorter-supporting lengths distribute too short to occur 
(1-1 and 1-2).  
Filter 1-1: P-values are less than the threshold_p(default: 10 ^ (-6)).  
Filter 1-2: The longest shorter-supporting lengths is shorter than 40% of the 
read length.  
Filter 2  : Hairpin-structure induced error detection (2-1 or 2-2).  
Filter 2-1: Palindromic sequences exist within 150 bases. 
Filter 2-2: >=50% mutation-supporting reads contains a reverse complementary 
sequence of the opposite strand consisting >= 15 bases.  
Filter 3  : 3’-/5’-supporting lengths are too densely distributed to occur 
(3-1, 3-2, and 3-3).  
Filter 3-1: P-values are less than the threshold_p(default: 10^(-6)).  
Filter 3-2: The distributions of 3’-/5’-supporting lengths are shorter than 80% 
of the read length.  
Filter 3-3: <10% of bases are low quality (Quality score <18).  
Filter 4  : >=90% mutation-supporting reads are soft-clipped (after cutting 
adapter sequence).  
Filter 5  : >=20% mutations were called by chimeric reads comprising two 
distant regions.  
Filter 6  : Mutations locating at simple repeat sequences.  
Filter 7  : C>T_g positive artifacts in FFPE samples.  
Filter 8  : Indel mutations locating at a >=15 homopolymer.  
  
Supporting lengths are adjusted considering small repeat sequences around the 
mutations.  
  
Results are saved in a excel file.  
The explatation of the results is written in detail in the second sheet of the 
excel file.  
  
github url: https://github.com/MANO-B/MicroSEC  
  
## Setting
```{r}
wd = "."
knitr::opts_chunk$set(collapse = TRUE,
                      fig.width = 12,
                      fig.height = 8,
                      echo = TRUE,
                      warning = FALSE,
                      message = TRUE,
                      comment = "#>")
options(rmarkdown.html_vignette.check_title = FALSE)
progress_bar = "N"
```

## Necessary packages
```{r packages}
library(MicroSEC)
```

## Analysis
```{r}
# initialize
msec = NULL
homology_search = NULL
mut_depth = NULL

# test data
sample_name = "H15-11943-1-T_TDv3"
read_length = 151
adapter_1 = "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA"
adapter_2 = "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"
organism = "hg38"

# load mutation information
data(exampleMutation)
data(exampleBAM)
data(exampleMutCall)

# load genomic sequence
ref_genome = fun_load_genome(organism)
chr_no = fun_load_chr_no(organism)

# analysis
result = fun_read_check(df_mutation = exampleMutation,
                        df_bam = exampleBAM,
                        df_mut_call = exampleMutCall,
                        ref_genome = ref_genome,
                        sample_name = sample_name,
                        read_length = read_length,
                        adapter_1 = adapter_1,
                        adapter_2 = adapter_2,
                        short_homology_search_length = 4,
                        progress_bar = progress_bar)
msec = rbind(msec, result[[1]])
homology_search = rbind(homology_search, result[[2]])
mut_depth = rbind(mut_depth, result[[3]])

# search homologous sequences
msec = fun_homology(msec,
                    homology_search,
                    min_homology_search = 40,
                    ref_genome,
                    chr_no,
                    progress_bar = progress_bar)

# statistical analysis
msec = fun_summary(msec)
msec = fun_analysis(msec,
                    mut_depth,
                    short_homology_search_length = 4,
                    min_homology_search = 40,
                    threshold_p = 10 ^ (-6),
                    threshold_hairpin_ratio = 0.50,
                    threshold_soft_clip_ratio = 0.90,
                    threshold_short_length = 0.8,
                    threshold_distant_homology = 0.2,
                    threshold_low_quality_rate = 0.1,
                    homopolymer_length = 15)

# save the results in the working/output directory.
# fun_save(msec, sample_name, wd)
```

## Results
```{r}
msec
```

## Information about the current R session
```{r}
sessionInfo()
```

