<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Masachika Ikegami" />

<meta name="date" content="2020-11-21" />

<title>MicroSEC: Sequence artifact filtering pipeline for FFPE samples</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>
<style type="text/css">
a.anchor-section {margin-left: 10px; visibility: hidden; color: inherit;}
a.anchor-section::before {content: '#';}
.hasAnchor:hover a.anchor-section {visibility: visible;}
</style>
<script>// Anchor sections v1.0 written by Atsushi Yasumoto on Oct 3rd, 2020.
document.addEventListener('DOMContentLoaded', function() {
  // Do nothing if AnchorJS is used
  if (typeof window.anchors === 'object' && anchors.hasOwnProperty('hasAnchorJSLink')) {
    return;
  }

  const h = document.querySelectorAll('h1, h2, h3, h4, h5, h6');

  // Do nothing if sections are already anchored
  if (Array.from(h).some(x => x.classList.contains('hasAnchor'))) {
    return null;
  }

  // Use section id when pandoc runs with --section-divs
  const section_id = function(x) {
    return ((x.classList.contains('section') || (x.tagName === 'SECTION'))
            ? x.id : '');
  };

  // Add anchors
  h.forEach(function(x) {
    const id = x.id || section_id(x.parentElement);
    if (id === '') {
      return null;
    }
    let anchor = document.createElement('a');
    anchor.href = '#' + id;
    anchor.classList = ['anchor-section'];
    x.classList.add('hasAnchor');
    x.appendChild(anchor);
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MicroSEC: Sequence artifact filtering pipeline for FFPE samples</h1>
<h4 class="author">Masachika Ikegami</h4>
<h4 class="date">2020-11-21</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This pipeline is designed for filtering mutations found in formalin-fixed and paraffin-embedded (FFPE) samples. The MicroSEC filter utilizes a statistical analysis, and the results for mutations with less than 10 supporting reads are not reliable. Four files are nessesary for the analysis: mutation information file, BAM file, and mutation supporting read ID information file.</p>
<p>File 1: mutation information file<br />
This excel file should contain at least these contents:</p>
<pre><code>   Sample     Gene HGVS.c  HGVS.p Mut_type Total_QV&gt;=20   %Alt  Chr  </code></pre>
<p>SL_1010-N6-B SLC25A24 _ _ 1-snv 366 1.0929 chr1</p>
<pre><code>  Pos Ref Alt SimpleRepeat_TRF                     Neighborhood_sequence  </code></pre>
<p>108130741 C T N CTACCTGGAGAATGGGCCCATGTGTCCAGGTAGCAGTAAGC</p>
<p>Transition<br />
C&gt;T_t</p>
<p>Total_QV&gt;=20: The read number with total Q-value &gt;=20.<br />
SimpleRepeat_TRF: Whether the mutation locates at a simple repeat sequence or not.<br />
Neighborhood_sequence: [5’-20 bases] + [Alt sequence] + [3’-20 bases].<br />
Transition: 1-snv mutation pattern with a 3’-base. C&gt;T_t represents CT to TT mutation. C&gt;T_g_FFPE represents the possible FFPE artifact.<br />
Sample, Mut_type, Chr, Pos, Ref, and Alt should be set exactly.<br />
Gene, HGVS.c, HGVS.p, Total_QV&gt;=20, %Alt, SimpleRepeat_TRF, and Transition can be set to any values.</p>
<p>File 2: BAM file</p>
<p>File 3: mutation supporting read ID information file<br />
This file should contain at least these contents:</p>
<p>Chr Pos Ref Alt<br />
chr1 2561609 T A</p>
<pre><code>                                                         Mut_ID     Mut  </code></pre>
<p>_;ID001-1:579185f,ID004-1:1873933f;ID006-1:1131647f,ID001-1:570086f .;A;N#</p>
<p>File 4: sample information tsv file<br />
Seven or eight columns are necessary.<br />
[sample name] [mutation information excel file] [BAM file] [read ID information directory] [read length] [adapter sequence read 1] [optional: adapter sequence read 2] [sample type: Human or Mouse]<br />
PC9 ./source/CCLE.xlsx ./source/Cell_line/PC9_Cell_line_Ag_TDv4.realigned.bam ./source/PC9_Cell_line 127 AGATCGGAAGAGCACACGTCTGAACTCCAGTCA AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT hg38</p>
<p>This pipeline contains 8 filtering processes.</p>
<p>Filter 1 : Shorter-supporting lengths distribute too short to occur (1-1 and 1-2).<br />
Filter 1-1: P-values are less than the threshold_p(default: 10 ^ (-6)).<br />
Filter 1-2: The longest shorter-supporting lengths is shorter than 40% of the read length.<br />
Filter 2 : Hairpin-structure induced error detection (2-1 or 2-2).<br />
Filter 2-1: Palindromic sequences exist within 150 bases. Filter 2-2: &gt;=50% mutation-supporting reads contains a reverse complementary sequence of the opposite strand consisting &gt;= 15 bases.<br />
Filter 3 : 3’-/5’-supporting lengths are too densely distributed to occur (3-1, 3-2, and 3-3).<br />
Filter 3-1: P-values are less than the threshold_p(default: 10^(-6)).<br />
Filter 3-2: The distributions of 3’-/5’-supporting lengths are shorter than 80% of the read length.<br />
Filter 3-3: &lt;10% of bases are low quality (Quality score &lt;18).<br />
Filter 4 : &gt;=90% mutation-supporting reads are soft-clipped (after cutting adapter sequence).<br />
Filter 5 : &gt;=20% mutations were called by chimeric reads comprising two distant regions.<br />
Filter 6 : Mutations locating at simple repeat sequences.<br />
Filter 7 : C&gt;T_g positive artifacts in FFPE samples.<br />
Filter 8 : Indel mutations locating at a &gt;=15 homopolymer.</p>
<p>Supporting lengths are adjusted considering small repeat sequences around the mutations.</p>
<p>Results are saved in a excel file.<br />
The explatation of the results is written in detail in the second sheet of the excel file.</p>
<p>github url: <a href="https://github.com/MANO-B/MicroSEC" class="uri">https://github.com/MANO-B/MicroSEC</a></p>
</div>
<div id="setting" class="section level2">
<h2>Setting</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>wd =<span class="st"> &quot;/mnt/HDD8TB/MicroSEC/archive&quot;</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>knitr<span class="op">::</span>opts_chunk<span class="op">$</span><span class="kw">set</span>(<span class="dt">collapse =</span> <span class="ot">TRUE</span>,</span>
<span id="cb4-3"><a href="#cb4-3"></a>                      <span class="dt">fig.width =</span> <span class="dv">12</span>,</span>
<span id="cb4-4"><a href="#cb4-4"></a>                      <span class="dt">fig.height =</span> <span class="dv">8</span>,</span>
<span id="cb4-5"><a href="#cb4-5"></a>                      <span class="dt">echo =</span> <span class="ot">TRUE</span>,</span>
<span id="cb4-6"><a href="#cb4-6"></a>                      <span class="dt">warning =</span> <span class="ot">FALSE</span>,</span>
<span id="cb4-7"><a href="#cb4-7"></a>                      <span class="dt">message =</span> <span class="ot">TRUE</span>,</span>
<span id="cb4-8"><a href="#cb4-8"></a>                      <span class="dt">comment =</span> <span class="st">&quot;#&gt;&quot;</span>)</span>
<span id="cb4-9"><a href="#cb4-9"></a>knitr<span class="op">::</span>opts_knit<span class="op">$</span><span class="kw">set</span>(<span class="dt">root.dir =</span> wd)</span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="kw">options</span>(<span class="dt">rmarkdown.html_vignette.check_title =</span> <span class="ot">FALSE</span>)</span>
<span id="cb4-11"><a href="#cb4-11"></a>progress_bar =<span class="st"> &quot;N&quot;</span></span></code></pre></div>
</div>
<div id="necessary-packages" class="section level2">
<h2>Necessary packages</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">library</span>(MicroSEC)</span></code></pre></div>
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># initialize</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>msec =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>homology_search =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>mut_depth =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb6-5"><a href="#cb6-5"></a></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co"># test data</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>sample_name =<span class="st"> &quot;H15-11943-1-T_TDv3&quot;</span></span>
<span id="cb6-8"><a href="#cb6-8"></a>read_length =<span class="st"> </span><span class="dv">151</span></span>
<span id="cb6-9"><a href="#cb6-9"></a>adapter_<span class="dv">1</span> =<span class="st"> &quot;AGATCGGAAGAGCACACGTCTGAACTCCAGTCA&quot;</span></span>
<span id="cb6-10"><a href="#cb6-10"></a>adapter_<span class="dv">2</span> =<span class="st"> &quot;AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT&quot;</span></span>
<span id="cb6-11"><a href="#cb6-11"></a>organism =<span class="st"> &quot;hg38&quot;</span></span>
<span id="cb6-12"><a href="#cb6-12"></a></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co"># load mutation information</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="kw">data</span>(exampleMutation)</span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="kw">data</span>(exampleBAM)</span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="kw">data</span>(exampleMutCall)</span>
<span id="cb6-17"><a href="#cb6-17"></a></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="co"># load genomic sequence</span></span>
<span id="cb6-19"><a href="#cb6-19"></a>ref_genome =<span class="st"> </span><span class="kw">fun_load_genome</span>(organism)</span>
<span id="cb6-20"><a href="#cb6-20"></a>chr_no =<span class="st"> </span><span class="kw">fun_load_chr_no</span>(organism)</span>
<span id="cb6-21"><a href="#cb6-21"></a></span>
<span id="cb6-22"><a href="#cb6-22"></a><span class="co"># analysis</span></span>
<span id="cb6-23"><a href="#cb6-23"></a>result =<span class="st"> </span><span class="kw">fun_read_check</span>(<span class="dt">df_mutation =</span> exampleMutation,</span>
<span id="cb6-24"><a href="#cb6-24"></a>                        <span class="dt">df_bam =</span> exampleBAM,</span>
<span id="cb6-25"><a href="#cb6-25"></a>                        <span class="dt">df_mut_call =</span> exampleMutCall,</span>
<span id="cb6-26"><a href="#cb6-26"></a>                        <span class="dt">ref_genome =</span> ref_genome,</span>
<span id="cb6-27"><a href="#cb6-27"></a>                        <span class="dt">sample_name =</span> sample_name,</span>
<span id="cb6-28"><a href="#cb6-28"></a>                        <span class="dt">read_length =</span> read_length,</span>
<span id="cb6-29"><a href="#cb6-29"></a>                        <span class="dt">adapter_1 =</span> adapter_<span class="dv">1</span>,</span>
<span id="cb6-30"><a href="#cb6-30"></a>                        <span class="dt">adapter_2 =</span> adapter_<span class="dv">2</span>,</span>
<span id="cb6-31"><a href="#cb6-31"></a>                        <span class="dt">short_homology_search_length =</span> <span class="dv">4</span>,</span>
<span id="cb6-32"><a href="#cb6-32"></a>                        <span class="dt">progress_bar =</span> progress_bar)</span>
<span id="cb6-33"><a href="#cb6-33"></a>msec =<span class="st"> </span><span class="kw">rbind</span>(msec, result[[<span class="dv">1</span>]])</span>
<span id="cb6-34"><a href="#cb6-34"></a>homology_search =<span class="st"> </span><span class="kw">rbind</span>(homology_search, result[[<span class="dv">2</span>]])</span>
<span id="cb6-35"><a href="#cb6-35"></a>mut_depth =<span class="st"> </span><span class="kw">rbind</span>(mut_depth, result[[<span class="dv">3</span>]])</span>
<span id="cb6-36"><a href="#cb6-36"></a></span>
<span id="cb6-37"><a href="#cb6-37"></a><span class="co"># search homologous sequences</span></span>
<span id="cb6-38"><a href="#cb6-38"></a>msec =<span class="st"> </span><span class="kw">fun_homology</span>(msec,</span>
<span id="cb6-39"><a href="#cb6-39"></a>                    homology_search,</span>
<span id="cb6-40"><a href="#cb6-40"></a>                    <span class="dt">min_homology_search =</span> <span class="dv">40</span>,</span>
<span id="cb6-41"><a href="#cb6-41"></a>                    ref_genome,</span>
<span id="cb6-42"><a href="#cb6-42"></a>                    chr_no,</span>
<span id="cb6-43"><a href="#cb6-43"></a>                    <span class="dt">progress_bar =</span> progress_bar)</span>
<span id="cb6-44"><a href="#cb6-44"></a></span>
<span id="cb6-45"><a href="#cb6-45"></a><span class="co"># statistical analysis</span></span>
<span id="cb6-46"><a href="#cb6-46"></a>msec =<span class="st"> </span><span class="kw">fun_summary</span>(msec)</span>
<span id="cb6-47"><a href="#cb6-47"></a>msec =<span class="st"> </span><span class="kw">fun_analysis</span>(msec,</span>
<span id="cb6-48"><a href="#cb6-48"></a>                    mut_depth,</span>
<span id="cb6-49"><a href="#cb6-49"></a>                    <span class="dt">short_homology_search_length =</span> <span class="dv">4</span>,</span>
<span id="cb6-50"><a href="#cb6-50"></a>                    <span class="dt">min_homology_search =</span> <span class="dv">40</span>,</span>
<span id="cb6-51"><a href="#cb6-51"></a>                    <span class="dt">threshold_p =</span> <span class="dv">10</span> <span class="op">^</span><span class="st"> </span>(<span class="op">-</span><span class="dv">6</span>),</span>
<span id="cb6-52"><a href="#cb6-52"></a>                    <span class="dt">threshold_hairpin_ratio =</span> <span class="fl">0.50</span>,</span>
<span id="cb6-53"><a href="#cb6-53"></a>                    <span class="dt">threshold_soft_clip_ratio =</span> <span class="fl">0.90</span>,</span>
<span id="cb6-54"><a href="#cb6-54"></a>                    <span class="dt">threshold_short_length =</span> <span class="fl">0.8</span>,</span>
<span id="cb6-55"><a href="#cb6-55"></a>                    <span class="dt">threshold_distant_homology =</span> <span class="fl">0.2</span>,</span>
<span id="cb6-56"><a href="#cb6-56"></a>                    <span class="dt">threshold_low_quality_rate =</span> <span class="fl">0.1</span>,</span>
<span id="cb6-57"><a href="#cb6-57"></a>                    <span class="dt">homopolymer_length =</span> <span class="dv">15</span>)</span>
<span id="cb6-58"><a href="#cb6-58"></a></span>
<span id="cb6-59"><a href="#cb6-59"></a><span class="co"># save the results</span></span>
<span id="cb6-60"><a href="#cb6-60"></a><span class="kw">fun_save</span>(msec, sample_name, wd)</span></code></pre></div>
</div>
<div id="results" class="section level2">
<h2>Results</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>msec</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#&gt;               Sample   Gene                 HGVS.c       HGVS.p Mut_type</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">#&gt; 1 H15-11943-1-T_TDv3  KMT2D              c.2887G&gt;A  p.Ala963Thr    1-snv</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; 2 H15-11943-1-T_TDv3  KMT2D              c.2886T&gt;C  p.Gly962Gly    1-snv</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">#&gt; 3 H15-11943-1-T_TDv3   TP53           c.783-114G&gt;A            _    1-snv</span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; 4 H15-11943-1-T_TDv3   TP53           c.783-117T&gt;C            _    1-snv</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">#&gt; 5 H15-11943-1-T_TDv3 PIK3CA         c.-76-3489delA            _    1-del</span></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co">#&gt; 6 H15-11943-1-T_TDv3 PIK3CA              c.3073A&gt;G p.Thr1025Ala    1-snv</span></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">#&gt; 7 H15-11943-1-T_TDv3   EGFR              c.2573T&gt;G  p.Leu858Arg    1-snv</span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">#&gt; 8 H15-11943-1-T_TDv3   EGFR c.3114+15_3114+16delTC            _    2-del</span></span>
<span id="cb7-11"><a href="#cb7-11"></a><span class="co">#&gt;   Total_QV&gt;=20     %Alt   Chr       Pos Ref Alt SimpleRepeat_TRF</span></span>
<span id="cb7-12"><a href="#cb7-12"></a><span class="co">#&gt; 1          163  6.74847 chr12  49050701   C   T                N</span></span>
<span id="cb7-13"><a href="#cb7-13"></a><span class="co">#&gt; 2          162  7.40741 chr12  49050702   A   G                N</span></span>
<span id="cb7-14"><a href="#cb7-14"></a><span class="co">#&gt; 3          189  8.46561 chr17   7673951   C   T                N</span></span>
<span id="cb7-15"><a href="#cb7-15"></a><span class="co">#&gt; 4          183  9.83607 chr17   7673954   A   G                N</span></span>
<span id="cb7-16"><a href="#cb7-16"></a><span class="co">#&gt; 5          134 17.16420  chr3 179195260  GA   G                N</span></span>
<span id="cb7-17"><a href="#cb7-17"></a><span class="co">#&gt; 6          185 27.56760  chr3 179234230   A   G                N</span></span>
<span id="cb7-18"><a href="#cb7-18"></a><span class="co">#&gt; 7          315 44.44440  chr7  55191822   T   G                N</span></span>
<span id="cb7-19"><a href="#cb7-19"></a><span class="co">#&gt; 8          231  6.92641  chr7  55201369 GTC   G                N</span></span>
<span id="cb7-20"><a href="#cb7-20"></a><span class="co">#&gt;                       Neighborhood_sequence Transition read_length total_read</span></span>
<span id="cb7-21"><a href="#cb7-21"></a><span class="co">#&gt; 1 AGGGTCACTGTCCCCTTTGGTACCAAAGGGGTACTCTAACT         NA         151         11</span></span>
<span id="cb7-22"><a href="#cb7-22"></a><span class="co">#&gt; 2 GGGTCACTGTCCCCTTTGGCGCCAAAGGGGTACTCTAACTC         NA         151         12</span></span>
<span id="cb7-23"><a href="#cb7-23"></a><span class="co">#&gt; 3 ACCCTTGTCCTTTCTGGAGCTTAAGCTCCAGCTCCAGGTAG         NA         151         16</span></span>
<span id="cb7-24"><a href="#cb7-24"></a><span class="co">#&gt; 4 CTTGTCCTTTCTGGAGCCTAGGCTCCAGCTCCAGGTAGGTG         NA         151         18</span></span>
<span id="cb7-25"><a href="#cb7-25"></a><span class="co">#&gt; 5 TATTTACTTGCCTTCTCGGTGAAAAAAAAAACAACAACAAA         NA         151         23</span></span>
<span id="cb7-26"><a href="#cb7-26"></a><span class="co">#&gt; 6 ACATTGCATACATTCGAAAGGCCCTAGCCTTAGATAAAACT         NA         151         51</span></span>
<span id="cb7-27"><a href="#cb7-27"></a><span class="co">#&gt; 7 CAAGATCACAGATTTTGGGCGGGCCAAACTGCTGGGTGCGG         NA         151        140</span></span>
<span id="cb7-28"><a href="#cb7-28"></a><span class="co">#&gt; 8 CTCTCTGGTATGAAATCTCTGTCTCTCTCTCTCTCAAGCTG         NA         151         16</span></span>
<span id="cb7-29"><a href="#cb7-29"></a><span class="co">#&gt;   soft_clipped_read flag_hairpin pre_support_length post_support_length</span></span>
<span id="cb7-30"><a href="#cb7-30"></a><span class="co">#&gt; 1                10           10                132                 137</span></span>
<span id="cb7-31"><a href="#cb7-31"></a><span class="co">#&gt; 2                11           10                133                 136</span></span>
<span id="cb7-32"><a href="#cb7-32"></a><span class="co">#&gt; 3                14           16                140                 134</span></span>
<span id="cb7-33"><a href="#cb7-33"></a><span class="co">#&gt; 4                14           16                143                 131</span></span>
<span id="cb7-34"><a href="#cb7-34"></a><span class="co">#&gt; 5                 9            0                130                 130</span></span>
<span id="cb7-35"><a href="#cb7-35"></a><span class="co">#&gt; 6                 7            0                144                 131</span></span>
<span id="cb7-36"><a href="#cb7-36"></a><span class="co">#&gt; 7                29            0                147                 135</span></span>
<span id="cb7-37"><a href="#cb7-37"></a><span class="co">#&gt; 8                 3            0                 97                 144</span></span>
<span id="cb7-38"><a href="#cb7-38"></a><span class="co">#&gt;   short_support_length low_quality_base_rate_under_q18 distant_homology_rate</span></span>
<span id="cb7-39"><a href="#cb7-39"></a><span class="co">#&gt; 1                   11                     0.059602649                     0</span></span>
<span id="cb7-40"><a href="#cb7-40"></a><span class="co">#&gt; 2                   10                     0.071743929                     0</span></span>
<span id="cb7-41"><a href="#cb7-41"></a><span class="co">#&gt; 3                   10                     0.007036424                     0</span></span>
<span id="cb7-42"><a href="#cb7-42"></a><span class="co">#&gt; 4                   18                     0.006990434                     0</span></span>
<span id="cb7-43"><a href="#cb7-43"></a><span class="co">#&gt; 5                   70                     0.039447164                     0</span></span>
<span id="cb7-44"><a href="#cb7-44"></a><span class="co">#&gt; 6                   72                     0.014673419                     0</span></span>
<span id="cb7-45"><a href="#cb7-45"></a><span class="co">#&gt; 7                   68                     0.006196783                     0</span></span>
<span id="cb7-46"><a href="#cb7-46"></a><span class="co">#&gt; 8                   75                     0.011175497                     0</span></span>
<span id="cb7-47"><a href="#cb7-47"></a><span class="co">#&gt;   prob_filter_1 prob_filter_3_pre prob_filter_3_post</span></span>
<span id="cb7-48"><a href="#cb7-48"></a><span class="co">#&gt; 1  2.794410e-11      0.5101794659       0.2681971521</span></span>
<span id="cb7-49"><a href="#cb7-49"></a><span class="co">#&gt; 2  5.248589e-17      0.4120721277       0.2379553368</span></span>
<span id="cb7-50"><a href="#cb7-50"></a><span class="co">#&gt; 3  2.254597e-19      0.1180670870       0.0519644737</span></span>
<span id="cb7-51"><a href="#cb7-51"></a><span class="co">#&gt; 4  7.602715e-14      0.1600788940       0.0635856288</span></span>
<span id="cb7-52"><a href="#cb7-52"></a><span class="co">#&gt; 5  1.384192e-03      0.0304539889       0.0161038299</span></span>
<span id="cb7-53"><a href="#cb7-53"></a><span class="co">#&gt; 6  1.134132e-01      0.0115221008       0.0157100141</span></span>
<span id="cb7-54"><a href="#cb7-54"></a><span class="co">#&gt; 7  8.203053e-06      0.0537795550       0.0001750834</span></span>
<span id="cb7-55"><a href="#cb7-55"></a><span class="co">#&gt; 8  1.000000e+00      0.0004783243       0.1146106607</span></span>
<span id="cb7-56"><a href="#cb7-56"></a><span class="co">#&gt;   filter_1_mutation_intra_hairpin_loop filter_2_hairpin_structure</span></span>
<span id="cb7-57"><a href="#cb7-57"></a><span class="co">#&gt; 1                                 TRUE                       TRUE</span></span>
<span id="cb7-58"><a href="#cb7-58"></a><span class="co">#&gt; 2                                 TRUE                       TRUE</span></span>
<span id="cb7-59"><a href="#cb7-59"></a><span class="co">#&gt; 3                                 TRUE                       TRUE</span></span>
<span id="cb7-60"><a href="#cb7-60"></a><span class="co">#&gt; 4                                 TRUE                       TRUE</span></span>
<span id="cb7-61"><a href="#cb7-61"></a><span class="co">#&gt; 5                                FALSE                      FALSE</span></span>
<span id="cb7-62"><a href="#cb7-62"></a><span class="co">#&gt; 6                                FALSE                      FALSE</span></span>
<span id="cb7-63"><a href="#cb7-63"></a><span class="co">#&gt; 7                                FALSE                      FALSE</span></span>
<span id="cb7-64"><a href="#cb7-64"></a><span class="co">#&gt; 8                                FALSE                      FALSE</span></span>
<span id="cb7-65"><a href="#cb7-65"></a><span class="co">#&gt;   filter_3_microhomology_induced_mutation filter_4_soft_clipping</span></span>
<span id="cb7-66"><a href="#cb7-66"></a><span class="co">#&gt; 1                                   FALSE                   TRUE</span></span>
<span id="cb7-67"><a href="#cb7-67"></a><span class="co">#&gt; 2                                   FALSE                   TRUE</span></span>
<span id="cb7-68"><a href="#cb7-68"></a><span class="co">#&gt; 3                                   FALSE                  FALSE</span></span>
<span id="cb7-69"><a href="#cb7-69"></a><span class="co">#&gt; 4                                   FALSE                  FALSE</span></span>
<span id="cb7-70"><a href="#cb7-70"></a><span class="co">#&gt; 5                                   FALSE                  FALSE</span></span>
<span id="cb7-71"><a href="#cb7-71"></a><span class="co">#&gt; 6                                   FALSE                  FALSE</span></span>
<span id="cb7-72"><a href="#cb7-72"></a><span class="co">#&gt; 7                                   FALSE                  FALSE</span></span>
<span id="cb7-73"><a href="#cb7-73"></a><span class="co">#&gt; 8                                   FALSE                  FALSE</span></span>
<span id="cb7-74"><a href="#cb7-74"></a><span class="co">#&gt;   filter_5_highly_homologous_region filter_6_simple_repeat</span></span>
<span id="cb7-75"><a href="#cb7-75"></a><span class="co">#&gt; 1                             FALSE                  FALSE</span></span>
<span id="cb7-76"><a href="#cb7-76"></a><span class="co">#&gt; 2                             FALSE                  FALSE</span></span>
<span id="cb7-77"><a href="#cb7-77"></a><span class="co">#&gt; 3                             FALSE                  FALSE</span></span>
<span id="cb7-78"><a href="#cb7-78"></a><span class="co">#&gt; 4                             FALSE                  FALSE</span></span>
<span id="cb7-79"><a href="#cb7-79"></a><span class="co">#&gt; 5                             FALSE                  FALSE</span></span>
<span id="cb7-80"><a href="#cb7-80"></a><span class="co">#&gt; 6                             FALSE                  FALSE</span></span>
<span id="cb7-81"><a href="#cb7-81"></a><span class="co">#&gt; 7                             FALSE                  FALSE</span></span>
<span id="cb7-82"><a href="#cb7-82"></a><span class="co">#&gt; 8                             FALSE                  FALSE</span></span>
<span id="cb7-83"><a href="#cb7-83"></a><span class="co">#&gt;   filter_7_c_to_t_artifact filter_8_mutation_at_homopolymer    msec_filter_1234</span></span>
<span id="cb7-84"><a href="#cb7-84"></a><span class="co">#&gt; 1                    FALSE                            FALSE Artifact suspicious</span></span>
<span id="cb7-85"><a href="#cb7-85"></a><span class="co">#&gt; 2                    FALSE                            FALSE Artifact suspicious</span></span>
<span id="cb7-86"><a href="#cb7-86"></a><span class="co">#&gt; 3                    FALSE                            FALSE Artifact suspicious</span></span>
<span id="cb7-87"><a href="#cb7-87"></a><span class="co">#&gt; 4                    FALSE                            FALSE Artifact suspicious</span></span>
<span id="cb7-88"><a href="#cb7-88"></a><span class="co">#&gt; 5                    FALSE                            FALSE                    </span></span>
<span id="cb7-89"><a href="#cb7-89"></a><span class="co">#&gt; 6                    FALSE                            FALSE                    </span></span>
<span id="cb7-90"><a href="#cb7-90"></a><span class="co">#&gt; 7                    FALSE                            FALSE                    </span></span>
<span id="cb7-91"><a href="#cb7-91"></a><span class="co">#&gt; 8                    FALSE                            FALSE                    </span></span>
<span id="cb7-92"><a href="#cb7-92"></a><span class="co">#&gt;     msec_filter_12345     msec_filter_all comment</span></span>
<span id="cb7-93"><a href="#cb7-93"></a><span class="co">#&gt; 1 Artifact suspicious Artifact suspicious        </span></span>
<span id="cb7-94"><a href="#cb7-94"></a><span class="co">#&gt; 2 Artifact suspicious Artifact suspicious        </span></span>
<span id="cb7-95"><a href="#cb7-95"></a><span class="co">#&gt; 3 Artifact suspicious Artifact suspicious        </span></span>
<span id="cb7-96"><a href="#cb7-96"></a><span class="co">#&gt; 4 Artifact suspicious Artifact suspicious        </span></span>
<span id="cb7-97"><a href="#cb7-97"></a><span class="co">#&gt; 5                                                </span></span>
<span id="cb7-98"><a href="#cb7-98"></a><span class="co">#&gt; 6                                                </span></span>
<span id="cb7-99"><a href="#cb7-99"></a><span class="co">#&gt; 7                                                </span></span>
<span id="cb7-100"><a href="#cb7-100"></a><span class="co">#&gt; 8</span></span></code></pre></div>
</div>
<div id="information-about-the-current-r-session" class="section level2">
<h2>Information about the current R session</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">sessionInfo</span>()</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; R version 3.6.3 (2020-02-29)</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">#&gt; Running under: Ubuntu 18.04.4 LTS</span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="co">#&gt; </span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="co">#&gt; </span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="co">#&gt; locale:</span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              </span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span id="cb8-15"><a href="#cb8-15"></a><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span id="cb8-16"><a href="#cb8-16"></a><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span id="cb8-17"><a href="#cb8-17"></a><span class="co">#&gt; </span></span>
<span id="cb8-18"><a href="#cb8-18"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb8-19"><a href="#cb8-19"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb8-20"><a href="#cb8-20"></a><span class="co">#&gt; </span></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb8-22"><a href="#cb8-22"></a><span class="co">#&gt; [1] MicroSEC_1.0.10</span></span>
<span id="cb8-23"><a href="#cb8-23"></a><span class="co">#&gt; </span></span>
<span id="cb8-24"><a href="#cb8-24"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb8-25"><a href="#cb8-25"></a><span class="co">#&gt;  [1] Biobase_2.46.0                    pkgload_1.1.0                    </span></span>
<span id="cb8-26"><a href="#cb8-26"></a><span class="co">#&gt;  [3] tidyr_1.1.2                       R.utils_2.10.1                   </span></span>
<span id="cb8-27"><a href="#cb8-27"></a><span class="co">#&gt;  [5] gtools_3.8.2                      assertthat_0.2.1                 </span></span>
<span id="cb8-28"><a href="#cb8-28"></a><span class="co">#&gt;  [7] stats4_3.6.3                      BSgenome_1.54.0                  </span></span>
<span id="cb8-29"><a href="#cb8-29"></a><span class="co">#&gt;  [9] GenomeInfoDbData_1.2.2            Rsamtools_2.2.3                  </span></span>
<span id="cb8-30"><a href="#cb8-30"></a><span class="co">#&gt; [11] yaml_2.2.1                        remotes_2.2.0                    </span></span>
<span id="cb8-31"><a href="#cb8-31"></a><span class="co">#&gt; [13] sessioninfo_1.1.1                 pillar_1.4.6                     </span></span>
<span id="cb8-32"><a href="#cb8-32"></a><span class="co">#&gt; [15] backports_1.1.10                  lattice_0.20-41                  </span></span>
<span id="cb8-33"><a href="#cb8-33"></a><span class="co">#&gt; [17] glue_1.4.2                        digest_0.6.27                    </span></span>
<span id="cb8-34"><a href="#cb8-34"></a><span class="co">#&gt; [19] GenomicRanges_1.38.0              XVector_0.26.0                   </span></span>
<span id="cb8-35"><a href="#cb8-35"></a><span class="co">#&gt; [21] htmltools_0.5.0                   Matrix_1.2-18                    </span></span>
<span id="cb8-36"><a href="#cb8-36"></a><span class="co">#&gt; [23] R.oo_1.24.0                       XML_4.0-0                        </span></span>
<span id="cb8-37"><a href="#cb8-37"></a><span class="co">#&gt; [25] pkgconfig_2.0.3                   devtools_2.3.2                   </span></span>
<span id="cb8-38"><a href="#cb8-38"></a><span class="co">#&gt; [27] zlibbioc_1.32.0                   purrr_0.3.4                      </span></span>
<span id="cb8-39"><a href="#cb8-39"></a><span class="co">#&gt; [29] processx_3.4.4                    openxlsx_4.2.3                   </span></span>
<span id="cb8-40"><a href="#cb8-40"></a><span class="co">#&gt; [31] BiocParallel_1.20.1               tibble_3.0.4                     </span></span>
<span id="cb8-41"><a href="#cb8-41"></a><span class="co">#&gt; [33] generics_0.0.2                    IRanges_2.20.2                   </span></span>
<span id="cb8-42"><a href="#cb8-42"></a><span class="co">#&gt; [35] usethis_1.6.3                     ellipsis_0.3.1                   </span></span>
<span id="cb8-43"><a href="#cb8-43"></a><span class="co">#&gt; [37] withr_2.3.0                       SummarizedExperiment_1.16.1      </span></span>
<span id="cb8-44"><a href="#cb8-44"></a><span class="co">#&gt; [39] BiocGenerics_0.32.0               cli_2.1.0                        </span></span>
<span id="cb8-45"><a href="#cb8-45"></a><span class="co">#&gt; [41] magrittr_1.5                      crayon_1.3.4                     </span></span>
<span id="cb8-46"><a href="#cb8-46"></a><span class="co">#&gt; [43] memoise_1.1.0                     evaluate_0.14                    </span></span>
<span id="cb8-47"><a href="#cb8-47"></a><span class="co">#&gt; [45] ps_1.4.0                          R.methodsS3_1.8.1                </span></span>
<span id="cb8-48"><a href="#cb8-48"></a><span class="co">#&gt; [47] fs_1.5.0                          fansi_0.4.1                      </span></span>
<span id="cb8-49"><a href="#cb8-49"></a><span class="co">#&gt; [49] xml2_1.3.2                        pkgbuild_1.1.0                   </span></span>
<span id="cb8-50"><a href="#cb8-50"></a><span class="co">#&gt; [51] tools_3.6.3                       data.table_1.13.2                </span></span>
<span id="cb8-51"><a href="#cb8-51"></a><span class="co">#&gt; [53] prettyunits_1.1.1                 lifecycle_0.2.0                  </span></span>
<span id="cb8-52"><a href="#cb8-52"></a><span class="co">#&gt; [55] matrixStats_0.57.0                stringr_1.4.0                    </span></span>
<span id="cb8-53"><a href="#cb8-53"></a><span class="co">#&gt; [57] S4Vectors_0.24.4                  zip_2.1.1                        </span></span>
<span id="cb8-54"><a href="#cb8-54"></a><span class="co">#&gt; [59] DelayedArray_0.12.3               callr_3.5.1                      </span></span>
<span id="cb8-55"><a href="#cb8-55"></a><span class="co">#&gt; [61] Biostrings_2.54.0                 compiler_3.6.3                   </span></span>
<span id="cb8-56"><a href="#cb8-56"></a><span class="co">#&gt; [63] GenomeInfoDb_1.22.1               rlang_0.4.8                      </span></span>
<span id="cb8-57"><a href="#cb8-57"></a><span class="co">#&gt; [65] grid_3.6.3                        RCurl_1.98-1.2                   </span></span>
<span id="cb8-58"><a href="#cb8-58"></a><span class="co">#&gt; [67] rstudioapi_0.11                   bitops_1.0-6                     </span></span>
<span id="cb8-59"><a href="#cb8-59"></a><span class="co">#&gt; [69] rmarkdown_2.5                     testthat_2.3.2                   </span></span>
<span id="cb8-60"><a href="#cb8-60"></a><span class="co">#&gt; [71] roxygen2_7.1.1                    R6_2.4.1                         </span></span>
<span id="cb8-61"><a href="#cb8-61"></a><span class="co">#&gt; [73] GenomicAlignments_1.22.1          rtracklayer_1.46.0               </span></span>
<span id="cb8-62"><a href="#cb8-62"></a><span class="co">#&gt; [75] knitr_1.30                        dplyr_1.0.2                      </span></span>
<span id="cb8-63"><a href="#cb8-63"></a><span class="co">#&gt; [77] commonmark_1.7                    rprojroot_1.3-2                  </span></span>
<span id="cb8-64"><a href="#cb8-64"></a><span class="co">#&gt; [79] desc_1.2.0                        stringi_1.5.3                    </span></span>
<span id="cb8-65"><a href="#cb8-65"></a><span class="co">#&gt; [81] parallel_3.6.3                    Rcpp_1.0.5                       </span></span>
<span id="cb8-66"><a href="#cb8-66"></a><span class="co">#&gt; [83] vctrs_0.3.4                       BSgenome.Hsapiens.UCSC.hg38_1.4.1</span></span>
<span id="cb8-67"><a href="#cb8-67"></a><span class="co">#&gt; [85] tidyselect_1.1.0                  xfun_0.18</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
